ExceptionHandling.md

```diff
@@ -0,0 +1,385 @@
+
+_**Try to throw before mutating an object state. Leave the objects’ state unaltered.**_
+
+### When to catch exceptions
+
+Generally, you should avoid catching exceptions where you cannot act on them.
+The diaper and the log-rethrow anti-pattern indicate a caught exception in the wrong place.
+There are some cases in which you want to catch exceptions.
+
+#### Catch and Resolve
+
+If you can resolve the exceptional state and return back to a normal state, you can catch the
+exception.
+However, be careful that you consider this decision wisely as it poses the question: If the
+exception can be resolved, is it really exceptional?
+If the answer is no, you should revisit your code and consider using a meaningful return value or
+whether to split your method with SRP instead of throwing an exception.
+
+#### Catch-Rethrow
+
+In some cases it can be useful to catch an exception, do something with it and rethrow that
+exception.
+
+1. **user-friendly message**: In case you want to wrap your exception in another exception with
+   information necessary to provide the user with a friendly and understandable message.
+2. **unit test a thrown exception**: In case you write a unit test that checks for an exception
+   being thrown. Use with care.
+3. **report debug information in new message**: Wrapping the exception in another exception with
+   useful debug information in the message can make debugging easier as the information will end up
+   in the logs.
+4. **get rid of checked exception**: As [discussed earlier](#checked-exceptions), we don't want the
+   abstraction leak of checked exceptions.
+
+#### Global exception handling
+
+So we never want to catch unwanted exceptions. How do we save our users from bombardment with stack
+traces?
+We learned that we should **never expose stack traces to users** as this is a security issue.
+So should I rather swallow exceptions and ignore
+the [diaper anti-pattern](#diaper-swallowing-exceptions)?
+
+As a developer, you want to feel safe and rest assured that all exceptions are handled somewhere.
+
+The solution is to create a safety net, catching all exceptions that occur. A global exception
+handler!
+Developers can throw exceptions whenever they want without needing to ensure that they are handled
+and without being afraid to expose details to users of your software.
+
+A global exception handler should sit between your software and the
+user. [In Spring applications this can be at the REST API level](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/ExceptionHandler.html).
+As Vaadin computes the view server-side, we can use its exception and error handling classes as
+global exception handlers;
+See [Routing Exception Handler in Vaadin](https://vaadin.com/docs/v23/routing/exceptions)
+and [Custom Exception Handler in Vaadin](https://vaadin.com/docs/v23/routing/exceptions#custom-exception-handlers).
+
+
+[//]: # (## FAQ)
```

Comment:
An additional information section or citing your sources could also be nice so we can inform ourselves further 👍